design/kafka-upgrade/README.md

```diff
@@ -0,0 +1,175 @@
+# Supporting Kafka upgrades
+
+* Status: **Under discussion**
+* Discussion: [GitHub PR](https://github.com/strimzi/strimzi-kafka-operator/pull/623)
+
+## Motivation
+
+The Strimzi Cluster Operator (CO) will need to support upgrading the Kafka 
+version used on a cluster-by-cluster basis. 
+(Note this is a different thing to supporting Operator upgrades).
+
+## Scope
+
+The expectation is that a given version of the operator will support some 
+defined set of Kafka versions and the user can selector which should be 
+used for their cluster.
+When the user upgrages to a later version supported by the CO, the CO will upgrade the cluster in a zero-downtime manner.
+
+Support for downgrading to an earlier version is not a priority.
+
+Upgrading Zookeeper is out of scope. We will assume that the new version of Kafka can use the existing Zookeeper version.
+
+## Upgrade in Kafka
+
+The upgrade procedure for Kafka is [documented here](https://kafka.apache.org/documentation/#upgrade).
+
+## Changes to the `Kafka` CRD
+
+Add `Kafka.spec.kafka.version` to control which version of Kafka to use.
+This corresponds to using a different image for Kafka broker pods.
```

Comment:
I had, of course, forgotten that we've already exposed the `image` directly in the CRD. If we retain that we would need a way to know what kafka version a given image was using. I can't see a good way of achieving that, tbh. So should we remove support for configuring `image` directly on the CR, and set the image using the version? Note, we could still support configuring the CO with different images via env vars as we do now, only those would now be version _templates_, parameterised on the `version` taken from the CR.

Reply:
Having thought some more about this, I think it's actually OK to have _both_ `version` and `image` (obviously it's up to the user to make sure the `image` they're using is actually a kafka image of the given kafka `version`). Retaining image means we retain flexibility. To avoid forcing the user to have an explicit `image` the CO can be configured with a mapping for version -> default image.